### PR TITLE
feat: support basic auth

### LINKED ISSUE #3069: Consider handling `BasicAuth` credentials in Provider
It'd be great to automatically handle `BasicAuth` details from URLs, extracting the credentials and injecting them via request headers—this way, users don't need to bother with this manipulation themselves:

 - `https://<username>:<password>@mainnet.fuel.network/v1/graphql`

@LuizAsFight kindly provided an excellent example of how he got around this.

basic-auth

### PR DESCRIPTION
**Pull request by nedsalk**
- Closes #3069 

# Release notes

In this release, we:
- Added support for URLs with `BasicAuth` credentials in `Provider`

# Checklist

- [x] All **changes** are **covered** by **tests** (or not applicable)
- [x] All **changes** are **documented** (or not applicable)
- [x] I **reviewed** the **entire PR** myself (preferably, on GH UI)
- [x] I **described** all **Breaking Changes** (or there's none)

### CODE REVIEW

**Review by petertonysmith94 — DISMISSED** [review]

File: packages/account/src/providers/provider.ts [hunk 1]
```diff
@@ -451,6 +451,29 @@ export default class Provider {
     }
   }
 
+  private static handleBasicAuth(url: string): [string, ProviderOptions['requestMiddleware']] {
```
Comment (arboleya) [inline packages/account/src/providers/provider.ts:454]: Rename to fit next suggestions.

```suggestion
  private static extractBasicAuth(url: string): [string, ProviderOptions['requestMiddleware']] {
```

File: packages/account/src/providers/provider.ts [hunk 2]
```diff
@@ -451,6 +451,29 @@ export default class Provider {
...
+
+    return [
+      urlNoBasicAuth,
+      (request) => {
+        request.headers ??= {};
+        (request.headers as Record<string, unknown>).Authorization = auth;
+
+        return request;
+      },
+    ];
```
Comment (arboleya) [inline packages/account/src/providers/provider.ts:474]: Removing side-effects, keeping it simple.

```suggestion
    return { url, auth };
```

File: packages/account/src/providers/provider.ts [hunk 3]
```diff
@@ -460,8 +483,17 @@ export default class Provider {
    * @returns A promise that resolves to a Provider instance.
    */
   static async create(url: string, options: ProviderOptions = {}): Promise<Provider> {
-    const provider = new Provider(url, options);
+    const [urlToUse, addBasicAuth] = this.handleBasicAuth(url);
```
Comment (arboleya) [inline packages/account/src/providers/provider.ts:486]: According to the previous suggestions:

```suggestion
    const { url: urlToUse, auth } = this.extractBasicAuth(url);
```

File: packages/account/src/providers/provider.ts [hunk 4]
```diff
@@ -460,8 +483,17 @@ export default class Provider {
...
+    const provider = new Provider(urlToUse, {
+      ...options,
+      requestMiddleware: async (request) => {
+        await addBasicAuth?.(request);
+        return options.requestMiddleware?.(request) ?? request;
+      },
```
Comment (arboleya) [inline packages/account/src/providers/provider.ts:492]: Injecting auth conditionally within the flow.

```suggestion
      requestMiddleware: async (request: <TypeHere>) => {
        if (auth) {
          request.headers ??= {};
          request.headers.Authorization = auth;
        }

        return options.requestMiddleware?.(request) ?? request;
      },
```

**Review by arboleya — CHANGES_REQUESTED** [review]
Left a couple of suggestions to simplify the flow.

Pseudo-code, pls double-check.

**Review by petertonysmith94 — APPROVED** [review]

**Review by danielbate — APPROVED** [review]

**Review by Torres-ssf — APPROVED** [review]

**Review by arboleya — APPROVED** [review]

**Pull request merged**